storey/aggregations.py

```diff
@@ -123,13 +126,41 @@ async def _do(self, event):
             if isinstance(self._emit_policy, EmitEveryEvent):
                 await self._emit_event(key, event)
             elif isinstance(self._emit_policy, EmitAfterMaxEvent):
-                self._events_in_batch[key] = self._events_in_batch.get(key, 0) + 1
-                if self._events_in_batch[key] == self._emit_policy.max_events:
+                if key in self._events_in_batch:
+                    self._events_in_batch[key]['counter'] = self._events_in_batch[key].get('counter') + 1
+                else:
+                    self._events_in_batch[key] = {}
+                    self._events_in_batch[key]['counter'] = 1
+                    self._events_in_batch[key]['time'] = time.monotonic()
+                self._events_in_batch[key]['event'] = event
+                if self._emit_policy.timeout_secs and self._timeout_task is None:
+                    self._timeout_task = asyncio.get_running_loop().create_task(self._sleep_and_emit())
+                if self._events_in_batch[key]['counter'] == self._emit_policy.max_events:
+                    if key == self._timeout_task_key and self._timeout_task and not self._timeout_task.cancelled():
+                        self._timeout_task.cancel()
+                        self._timeout_task = None
+                        self._timeout_task_key = None
                     await self._emit_event(key, event)
-                    self._events_in_batch[key] = 0
+                    self._events_in_batch.pop(key)
+
         except Exception as ex:
             raise ex
 
+    async def _sleep_and_emit(self):
+        while self._events_in_batch:
+            key = next(iter(self._events_in_batch.keys()))
+            if self._events_in_batch[key]['event'] is None:
+                continue
```

Comment:
Can't this get stuck in an infinite loop?

Reply:
i don;t see a scenario where self._events_in_batch[key]['event'] is None. so perhaps it's better to delete this if altogether...